File: 2021/07/javascript/main.js
const fs = require('fs')

fs.readFile('input.txt', {encoding: 'utf-8'}, (_, input) => {
  const submarines = input.split(',').map(submarine => Number(submarine))
  const min = Math.min(...submarines)
  const max = Math.max(...submarines)
  const range = [...Array(max - min).keys()].map(value => value + min)

  console.log(
    'Lowest with steady fuel costs:',
    Math.min(...range.map(position => submarines.reduce((acc, submarine) => acc + Math.abs(submarine - position), 0))),
  )

  console.log(
    'Lowest with increasing fuel costs:',
    Math.min(...range.map(position => submarines.reduce((acc, submarine) => {
      const distance = Math.abs(submarine - position)
      return acc + (Math.pow(distance, 2) + distance) / 2
    }, 0))),
  )
})

File: 2021/07/rust/src/main.rs
use std::fs;

fn main() {
  let content = fs::read_to_string("input.txt").expect("Failed to read input file");
  let crap_submarines: Vec<i32> = content
    .split(",")
    .map(|submarine| submarine.parse::<i32>().expect("Failed to convert to integer"))
    .collect();

  let min = crap_submarines.iter().min().unwrap();
  let max = crap_submarines.iter().max().unwrap();

  println!(
    "Lowest with steady fuel costs: {}",
    (*min..=*max)
      .map(|position| crap_submarines
        .iter()
        .fold(0, |acc, submarine| acc + i32::abs(submarine - position)))
      .min()
      .unwrap()
  );

  println!(
    "Lowest with increasing fuel costs: {}",
    (*min..=*max)
      .map(|position| crap_submarines.iter().fold(0, |acc, submarine| {
        let distance = i32::abs(submarine - position);
        acc + ((distance.pow(2) + distance) / 2)
      }))
      .min()
      .unwrap()
  );
}
